Proactively drop closed subscribers while waiting slot/block import acknowledgements in `MockPrimaryNode`

NingLin-P opened the pull request:
Fix #1352 

It is possible that the slot/block import subscriber is dropped after notification then we will block on waiting for acknowledgment forever, see https://github.com/subspace/subspace/issues/1352#issuecomment-1499292448 for more details. This commit fixes this issue by proactively drop closed subscribers while waiting for acknowledgements.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

nazar-pc commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -184,8 +184,21 @@ impl MockPrimaryNode {
+                    expected_ack = self.new_slot_notification_subscribers.len();
+                }
+            }
         }
```
This feels much more complex than it should be. I think this is doing the same:
```suggestion
        // Wait for all the acknowledgements to progress and proactively drop closed subscribers.
        loop {
            select! {
                res = slot_acknowledgement_receiver.next() => if res.is_none() {
                    break;
                },
                // TODO: Workaround for https://github.com/smol-rs/async-channel/issues/23, remove once fix is released
                _ = futures_timer::Delay::new(time::Duration::from_millis(500)).fuse() => {
                    self.new_slot_notification_subscribers.retain(|subscriber| !subscriber.is_closed());
                }
            }
        }
```

NingLin-P replied:
You're right, thanks!

nazar-pc reviewed (commented):
Make sure to note what things are necessary and what are workarounds. The code you wrote here is an annoying workaround we wouldn't write otherwise, but there was no TODO to remove it, no link for prosperity about such monstrosity is needed. We want to leave instructions for future maintainers explaining when code can be cleaned up.

NingLin-P committed:
Proactively drop closed subscribers while waiting slot/block import acknowledgements in MockPrimaryNode

It is possible that the slot/block import subscriber is dropped after notification then
we will block on waiting for acknowledgement forever, see #1352 for more details. This
commit fix this issue by proactively drop closed subscribers while waiting acknowledgements.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Enable test test_executor_full_node_catching_up and invalid_execution_proof_should_not_work

Signed-off-by: linning <linningde25@gmail.com>
